File: src/TypeFormatter/DefinitionTypeFormatter.ts
import type { Definition } from "../Schema/Definition.js";
import type { SubTypeFormatter } from "../SubTypeFormatter.js";
import type { BaseType } from "../Type/BaseType.js";
import { DefinitionType } from "../Type/DefinitionType.js";
import type { TypeFormatter } from "../TypeFormatter.js";
import { uniqueArray } from "../Utils/uniqueArray.js";

export class DefinitionTypeFormatter implements SubTypeFormatter {
    public constructor(
        protected childTypeFormatter: TypeFormatter,
        protected encodeRefs: boolean,
    ) {}

    public supportsType(type: BaseType): boolean {
        return type instanceof DefinitionType;
    }
    public getDefinition(type: DefinitionType): Definition {
        const ref = type.getName();
        return { $ref: `#/definitions/${this.encodeRefs ? encodeURIComponent(ref) : ref}` };
    }
    public getChildren(type: DefinitionType): BaseType[] {
        return uniqueArray([type, ...this.childTypeFormatter.getChildren(type.getType())]);
    }
}

File: internal/formatter/definition_type_formatter.go
package formatter

import (
	"github.com/vega/ts-json-schema-generator/internal/schema"
	"github.com/vega/ts-json-schema-generator/internal/types"
)

// DefinitionTypeFormatter mirrors src/TypeFormatter/DefinitionTypeFormatter.ts.
type DefinitionTypeFormatter struct {
	childTypeFormatter TypeFormatter
	encodeRefs         bool
}

func NewDefinitionTypeFormatter(childTypeFormatter TypeFormatter, encodeRefs bool) *DefinitionTypeFormatter {
	return &DefinitionTypeFormatter{childTypeFormatter: childTypeFormatter, encodeRefs: encodeRefs}
}

func (f *DefinitionTypeFormatter) SupportsType(t types.Type) bool {
	_, ok := t.(*types.DefinitionType)
	return ok
}

func (f *DefinitionTypeFormatter) GetDefinition(t types.Type) *schema.Definition {
	ref := t.(*types.DefinitionType).Name()
	if f.encodeRefs {
		ref = schema.EncodeRef(ref)
	}
	return &schema.Definition{Ref: "#/definitions/" + ref}
}

func (f *DefinitionTypeFormatter) GetChildren(t types.Type) []types.Type {
	definitionType := t.(*types.DefinitionType)
	children := append([]types.Type{definitionType}, f.childTypeFormatter.GetChildren(definitionType.Type)...)
	return uniqueTypes(children)
}
